src/LuaUtils.cpp

```diff
@@ -171,127 +171,363 @@ static bool _import_core(lua_State *L, const std::string &importName)
 
 #undef DEBUG_IMPORT
 #ifdef DEBUG_IMPORT
-#define DEBUG_PRINTF Output
+	#define DEBUG_PRINTF Output
+	#define DEBUG_INDENTED_PRINTF IndentedOutput
+	#define DEBUG_INDENT_INCREASE IndentIncrease
+	#define DEBUG_INDENT_DECREASE IndentDecrease
 #else
-#define DEBUG_PRINTF(...)
+	#define DEBUG_PRINTF(...)
+	#define DEBUG_INDENTED_PRINTF(...)
+	#define DEBUG_INDENT_INCREASE(...)
+	#define DEBUG_INDENT_DECREASE(...)
 #endif
 
-static bool _import(lua_State *L, const std::string &importName)
+// simple struct to pass data between functions
+struct ImportInfo
 {
+	const std::string& importName; // original name argument from "import(importName)"
+	std::string& fileName; // name of the existing file
+	const bool& isFullName; // if true import will no try to load relative to the importDirectories
+	std::vector<std::string>& importDirectories; // contans list of paths where file can be imported
+};
+
+/**
+* Returns path of last element in the lua stack
+*/
+static std::string get_caller(lua_State *L)
+{
+	assert(L);
 	LUA_DEBUG_START(L);
 
-	DEBUG_PRINTF("import: request for %s\n", importName.c_str());
+	std::string caller;
+	lua_Debug ar;
 
-	// get the cache of previously imported things
-	lua_getfield(L, LUA_REGISTRYINDEX, "Imports");
+	// if have stack and caller in the stack
+	if (lua_getstack(L, 1, &ar) && lua_getinfo(L, "S", &ar) && ar.source)
+	{
+		int start = 0;
+		int end = strlen(ar.source);
+
+		if (ar.source[start] != '@' && end > 5)
+			start = 4;
+
+		if (ar.source[start] == '@') {
+			start++;
+
+			static char source[1024];
+			strncpy(source, &ar.source[start], end - start + 1);
+
+			caller = std::string(source);
```

Comment:
Why copy the string into an array if it's to wrap it in a std::string afterwards ? Couldn't you simply do
```cpp
caller = std::string(ar.source + start, end - start +1); // the +1 needs to be checked against the exact spec of std::string(char* src, size_t len);
```